python/dgl/transform/module.py

```diff
@@ -0,0 +1,750 @@
+#
+"""Modules for transform"""
+
+from .. import convert
+from .. import backend as F
+from . import functional
+
+__all__ = [
+    'BaseTransform',
+    'AddSelfLoop',
+    'RemoveSelfLoop',
+    'AddReverse',
+    'ToSimple',
+    'LineGraph',
+    'KHopGraph',
+    'AddMetaPaths',
+    'KNNGraph',
+    'Compose'
+]
+
+class BaseTransform:
+    r"""
+
+    Description
+    -----------
+    An abstract class for writing transforms.
+    """
+    def transform_symmetric_etype(self, c_etype, *args):
+        r"""
+
+        Description
+        -----------
+        Transform a relation graph whose source and destination node types are identical.
+
+        Parameters
+        ----------
+        c_etype : 3-tuple of str
+            A canonical edge type.
+        """
+        raise NotImplementedError
+
+    def transform_asymmetric_etype(self, c_etype, *args):
+        r"""
+
+        Description
+        -----------
+        Transform a relation graph whose source and destination node types are distinct.
+
+        Parameters
+        ----------
+        c_etype : 3-tuple of str
+            A canonical edge type.
+        """
+        raise NotImplementedError
+
+    def transform_etype(self, c_etype, *args):
+        r"""
+
+        Description
+        -----------
+        Transform a relation graph.
+
+        Parameters
+        ----------
+        c_etype : 3-tuple of str
+            A canonical edge type.
+        """
+        utype, _, vtype = c_etype
+        if utype == vtype:
+            return self.transform_symmetric_etype(c_etype, *args)
+        else:
+            return self.transform_asymmetric_etype(c_etype, *args)
+
+    def update_graph_structure(self, g, data_dict, copy_edata=True):
+        r"""
+
+        Description
+        -----------
+        Update the structure of a graph.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph to update.
+        data_dict : graph data
+            The dictionary data for constructing a heterogeneous graph.
+        copy_edata : bool
+            If True, it will copy the edge features to the updated graph.
+
+        Returns
+        -------
+        DGLGraph
+            The updated graph.
+        """
+        device = g.device
+        idtype = g.idtype
+        num_nodes_dict = dict()
+
+        for ntype in g.ntypes:
+            num_nodes_dict[ntype] = g.num_nodes(ntype)
+
+        new_g = convert.heterograph(data_dict, num_nodes_dict=num_nodes_dict,
+                                    idtype=idtype, device=device)
+
+        # Copy features
+        for ntype in g.ntypes:
+            for key, feat in g.nodes[ntype].data.items():
+                new_g.nodes[ntype].data[key] = feat
+
+        if copy_edata:
+            for c_etype in g.canonical_etypes:
+                for key, feat in g.edges[c_etype].data.items():
+                    new_g.edges[c_etype].data[key] = feat
+
+        return new_g
+
+    def __call__(self, g):
```

Comment:
I feel we only need `__call__` and `__repr__`. Could you explain why we need the other methods?